From aa2dd29211788f2d4b6a6588f8fd05b8518a9bc2 Mon Sep 17 00:00:00 2001
From: Alessandro Decina <alessandro.decina@collabora.co.uk>
Date: Tue, 6 Sep 2011 17:15:28 +0200
Subject: [PATCH 18/19] v4l2sink: v4l2-gfx driver cropping workaround

When using the v4l2-gfx driver, the rendering side usually
gets frame width and height looking at the
v4l2_gfx_buf_params structure passed to the V4L2_GFX_IOC_ACQ
ioctl. The structure is populated with the values set by the
last VIDIOC_S_CROP ioctl, or zero filled if VIDIOC_S_CROP
was never called.

So we make sure to call VIDIOC_S_CROP at least once, and
always with crop_width and crop_height set to meaningful
values.
---
 sys/v4l2/gstv4l2sink.c |    9 +++++----
 1 files changed, 5 insertions(+), 4 deletions(-)

diff --git a/sys/v4l2/gstv4l2sink.c b/sys/v4l2/gstv4l2sink.c
index 146803c..934c591 100644
--- a/sys/v4l2/gstv4l2sink.c
+++ b/sys/v4l2/gstv4l2sink.c
@@ -521,9 +521,6 @@ gst_v4l2sink_sync_overlay_fields (GstV4l2Sink * v4l2sink)
 static void
 gst_v4l2sink_sync_crop_fields (GstV4l2Sink * v4l2sink)
 {
-  if (!v4l2sink->crop_fields_set)
-    return;
-
   if (GST_V4L2_IS_OPEN (v4l2sink->v4l2object)) {
 
     gint fd = v4l2sink->v4l2object->video_fd;
@@ -537,12 +534,16 @@ gst_v4l2sink_sync_crop_fields (GstV4l2Sink * v4l2sink)
       return;
     }
 
-    GST_DEBUG_OBJECT (v4l2sink,
+    GST_INFO_OBJECT (v4l2sink,
         "setting crop: crop_fields_set=0x%02x, top=%d, left=%d, width=%d, height=%d",
         v4l2sink->crop_fields_set,
         v4l2sink->crop.top, v4l2sink->crop.left,
         v4l2sink->crop.width, v4l2sink->crop.height);
 
+    if (v4l2sink->crop_fields_set == 0) {
+      crop.c.width = GST_VIDEO_SINK_WIDTH (v4l2sink);
+      crop.c.height = GST_VIDEO_SINK_HEIGHT (v4l2sink);
+    }
     if (v4l2sink->crop_fields_set & RECT_TOP_SET)
       crop.c.top = v4l2sink->crop.top;
     if (v4l2sink->crop_fields_set & RECT_LEFT_SET)
-- 
1.7.1

